### DISCUSSION TITLE
How do I use useFieldArray with a generic Control Type?

### DISCUSSION THREAD

**Discussion by HvZimmermann** [react-hook-form/react-hook-form#11708] [Q&A] [answered]
I am trying to build a custom component that takes a generic control from a useForm. In this component i want to use useFieldArray to add a variable amount of input fields.

When I set my generic component up to take any fields additionally to the ones I care about, I always get a type error in various places along the lines of:

```
 The expected type comes from property 'name' which is declared here on type 'UseFieldArrayProps<T, ArrayPath<T>, "id">'
```

I want to do this, so that I can reuse my custom components with different forms that all include my relevant fields. 

Check my mini example here: https://codesandbox.io/p/sandbox/bold-hooks-pwq6k7?file=%2Fsrc%2FApp.tsx

**Comment by sujinleeme** [comment, marked as answer]
I grabbed the hint from [here](https://github.com/orgs/react-hook-form/discussions/9789).


[[playground]](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKFEljgG84BhCAOxhwBsAaOAVwDOSAGLAknACYBBKFBQBPXoJHQ8AXzjZccAERF0MALQALCBADWhzKp0VKVGPLBI4AcSSskUYGmGqAIigwKHAAvAxUcHAwxiJikgCSTChEUgBGEHwwAFwMcKwoIEi5AhzArADmcGoA2gC6dmp2Dk4uolClflAgAGoonHxIAmERlFHBAuYlZZWNcABkbh5ePl0ggcHNlI7OcADKSGhsEmt9A0Mj7p7evgFBKFs7LlcraCzgbB4wAAo4YAIAHgAKqd+oMBAA+Eb0SJwI7sLi5EGqM7guBIAAeMA8EmGLxuaw2IQA-Mw2BwIJxgaDzpC4LlPAA3LyNLaYPisDDANhLa4+d6QTzsakosEXTHY1i43mvQn3CEAChhYzh5K43Coaly+P5uEFX1+EH+Iu6qKGEIAlKMovDSnkUGBnFLqiNlKJxNJZAolbCbWrKRqVVECkVcjoYnEPUkUkh0pkYDpA1E1Batn7WHa2FJHTiAIrgmDc1iuHB8MAjBVW0IQ31wB1OiQ+oNRfKFYq6RO1lN2KJEGB8KCsOAAiTARkQgEAelH48aVComNo8AkSEwKD4nCwHK5POzYEr1tVGfgjHhFM4LvCbtUAPanVFtMVqdhtpPR-PJUOxzWLARlMv-BCGsAIHPCJwPuCT5pnAfYDkOCq1iOY41s2UQAjqbx6p87DvlwoT0GeXAaJOKEti26HLDcArYfAhGUvhQhgT+-qcMREI9uR07IbCz6UGoVBAA)

```tsx
import React from 'react';
import { Control, useFieldArray, useForm } from "react-hook-form";

type GenericFormData = {
  theFieldICareAbout: { name: string }[];
};

type FirstFormValues = {
  task: string;
} & GenericFormData;

type SecondFormValues = GenericFormData;

type GenericComponentProps<TFormValues> = {
  control: TFormValues extends GenericFormData ? Control<TFormValues> : never;
};

function GenericComponent<TFormValues extends GenericFormData>({
  control,
}: GenericComponentProps<TFormValues>) {
  const { append } = useFieldArray({
    control,
    name: "theFieldICareAbout",
  });

  const onAppendQuestionGroup = () =>
    append({
      name: "",
    });
  return <div></div>;
}

export default function App() {
  const { control } = useForm<FirstFormValues>();
  const { control: secondFormControl } = useForm<SecondFormValues>();

  return (
    <div>
      <GenericComponent control={control} />
      <GenericComponent control={secondFormControl} />;
    </div>
  );
}
```

**Reply by HvZimmermann** [reply]
Thank you!

**Reply by Blakeinstein** [reply]
Is it possible to also make the key `theFieldICareAbout` generic? Like the parent component can pass the key name as prop?

**Reply by mhuggins** [reply]
@Blakeinstein I'm struggling with the same thing.  Did you ever figure this out?

**Reply by Blakeinstein** [reply]
> @Blakeinstein I'm struggling with the same thing.  Did you ever figure this out?

Unfortunately no I dont think I was able to do that :(

**Reply by alyyousuf7** [reply]
I think I was able to come up with a solution - but it's not perfect:

```tsx
// ListDocumentInput.tsx
import { Card, List } from '@mui/material';
import { FieldArrayPath, useFieldArray, UseFieldArrayProps } from 'react-hook-form';

import ListItemDocument from './ListItemDocument';
import { Document } from './services';

type FieldValuesWith<TValue> = {
  [key: string]: TValue;
};

type UseFieldArrayWithGenericValueProps<TValue> = UseFieldArrayProps<
  FieldValuesWith<TValue[]>,
  FieldArrayPath<FieldValuesWith<TValue[]>>
>;

const ListDocumentInput = ({
  ...controllerProps
}: UseFieldArrayWithGenericValueProps<Document>) => {
  const { fields, remove } = useFieldArray({
    ...controllerProps,
    keyName: 'internalId',
  });

  return (
    <List component={Card}>
      {fields.map((doc, index) => (
        <ListItemDocument key={doc.internalId} document={doc} onRemove={() => remove(index)} />
      ))}
    </List>
  );
};

export default ListDocumentInput;
```

Usage:
```tsx
// App.tsx
type FormValues = {
  documents1: Documents[];
  documents2: Documents[];
};

export default function App() {
  const formMethods = useForm<FormValues>({ defaultValues: { documents: [] } });

  return (
    <FormProvider {...formMethods}>
      <ListDocumentInput name="documents1" />
      <ListDocumentInput name="documents2" />
    </FormProvider>
  );
}
```

It enforces the `FormValue` inside `ListDocumentInput.tsx` to consider all keys to have the generic type array value - hence making sure we get typings for it from `useFieldArray`
